[torch.compile] PostGradPassManager, Inductor code caching fix, fix_functionalization pass refactor + tests

ProExpertProg opened the pull request:
This PR extracts the post_grad passes into `PostGradPassManager`. By overriding pickling of the class, we avoid failing to pickle custom passes, which is used in the code cache.

It also extracts pass config into `PassConfig` and adds an option to disable the reshapes pass. It adds utilities for registering custom inductor passes, and extracts fix_functionalization into a proper pass.

ProExpertProg committed:
Refactor fix_functionalization into pass

Signed-off-by: luka <luka@neuralmagic.com>
Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

ProExpertProg committed:
Move is_func to inductor_pass.py

Signed-off-by: luka <luka@neuralmagic.com>
Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

ProExpertProg committed:
Final functionalization.py refactor - each case is now as low as 2 lines

Signed-off-by: luka <luka@neuralmagic.com>
Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

ProExpertProg committed:
Add test for fix_functionalization

Signed-off-by: luka <luka@neuralmagic.com>
Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

ProExpertProg committed:
Extract post grad passes into manager. Handles passes, config, pickling.

Add config option to disable reshape pass.

Signed-off-by: luka <luka@neuralmagic.com>
Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

ProExpertProg commented on vllm/compilation/backends.py:
```diff
@@ -362,6 +172,76 @@ def call_module(self, target: torch.fx.node.Target,
+        self.passes.append(pass_)
+        self.has_custom_passes = True
+
+    def __getstate__(self):
```
I tried that but it isn't, due to the fusion pass. But we could just ignore that one.

ProExpertProg replied:
I pivoted to a uuid-based serialization approach, like in torch==2.6.

ProExpertProg committed:
Pickling strategy following the torch==2.6 uuid approach

Signed-off-by: luka <luka@neuralmagic.com>
Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

ProExpertProg committed:
Move pass manager to own file

Signed-off-by: luka <luka@neuralmagic.com>
Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

ProExpertProg committed:
Add PassManager caching test

Signed-off-by: luka <luka@neuralmagic.com>
Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

ProExpertProg committed:
Move config into tests, explicitly disable caching in TestBackend

Signed-off-by: luka <luka@neuralmagic.com>
Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

ProExpertProg committed:
Add logging to fix_functionalization

Signed-off-by: luka <luka@neuralmagic.com>
Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

ProExpertProg committed:
Fix PassConfig docstring

Signed-off-by: luka <luka@neuralmagic.com>
Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

SageMoore commented on vllm/compilation/backends.py:
```diff
@@ -401,32 +205,28 @@ def __init__(self, post_grad_passes: Sequence[Callable] = ()):
-        self.post_grad_passes = post_grad_passes
+
+        # Passes to run on the graph post-grad.
+        self.post_grad_pass_manager = PostGradPassManager(
```
What passes are expected to be passed in at __init__ time vs __call__ time?

ProExpertProg replied:
This just allows anyone manually constructing `VllmBackend` to add additional passes. Generally known passes will depend on config and should be done at call time. I can add a comment

ProExpertProg replied:
This was removed

bnellnm commented on vllm/compilation/config.py:
```diff
@@ -6,12 +6,38 @@
+    enable_reshape: bool = True
+
+    def model_post_init(self, __context: Any) -> None:
+        if not self.enable_reshape and self.enable_fusion:
+            print_warning_once(
+                "Fusion enabled but reshape elimination disabled."
+                "RMSNorm + quant (fp8) fusion might not work")
```
Is there some reason why you would enable fusion without enabling the reshape pass?  Or just enabling the reshape pass alone?  i.e. why not one flag for both passes (or just call reshape at the end of fusion)

ProExpertProg replied:
Only fp8 fusion needs the reshape pass, int8 doesn't. I was specifically asked to allow disabling the reshape pass but don't feel strongly either way - but if we don't have them decoupled, we should disable fusion if reshape is disabled.

bnellnm replied:
I don't have strong feelings either but seems like the two passes work in conjunction and removing redundant reshapes would always be beneficial.

WoosukKwon replied:
@ProExpertProg @bnellnm Because of the compilation time. I found that it increased the compilation time while reshaping 0 ops for bf16 llama.

bnellnm replied:
The reshape pass looks dead simple.  Just a linear loop over nodes.  I can't believe it is any more expensive than fix_functionalization or fusion.

ProExpertProg replied:
@bnellnm I agree, so I've been looking into the compilation time. I assume its impact will be negligible, and it's actually the fusion pass and fix_functionalization that are expensive. But I'll report what the compilation times are here

ProExpertProg replied:
Timings are logged (under debug), and there's an option to disable the reshape pass

bnellnm commented on vllm/compilation/functionalization.py:
nit: I think the name of this file is a bit confusing (especially since the fix_functionalization pass selectively defunctionalizes certain ops).

ProExpertProg replied:
Didn't want it to be too long but agreed, I'll rename

bnellnm commented on vllm/compilation/inductor_pass.py:
```diff
@@ -14,13 +19,99 @@
+InductorPassType: TypeAlias = Union[InductorPass, Callable[[fx.Graph], None]]
+
+
+class CallableInductorPass(InductorPass):
```
why differentiate a `CallableInductorPass` from `InductorPass`?  Seems like all `InductorPasses` should be callable?

bnellnm commented on vllm/compilation/inductor_pass.py:
```diff
@@ -14,13 +19,99 @@
+        return self._uuid
+
+
+def as_inductor_pass(*, uuid: Any = None, files: Tuple[str, ...] = ()):
```
It looks like this function (and CallableInductorPass) is only used test code.  Can this code just live in the test directory?

ProExpertProg replied:
These functions are provided as utilities for vLLM users to add their own inductor passes and still work with the caching mechanism. I can add better comments here

bnellnm replied:
Couldn't they just inherit from `InductorPass` or whatever the appropriate class is?

ProExpertProg replied:
Yeah this just makes it easier, one can just slap an annotation on a function instead of rewriting it to a class. But happy to remove it if you think I should. @youkaichao would love your input on what the best way to expose custom passes to users would be.

youkaichao replied:
I think using uuid for passes is quite hacky. can we use function qualname instead?

youkaichao commented on vllm/compilation/pass_manager.py:
```diff
@@ -0,0 +1,82 @@
+
+        self.fix_functionalization = FixFunctionalizationPass(pass_config)
+
+    def add(self, pass_: InductorPassType):
```
I think by using qualified name for the passes, we can avoid the serialization problem.

youkaichao replied:
One downside of using qualified name, is it does not support partially bound functions, e.g. if a pass is something like `partial(general_transform_func, args)` . I don't know if we have this kind of use cases right now.

ProExpertProg replied:
The issue with that is that if a user changes the implementation for a function, the qualified name stays the same and the inductor cache will use the outdated code, even though it should recompile with the new function implementation.

youkaichao replied:
I think that is fine. that's only for developers. and we can issue a warning to tell developers that they should always recompile when they are developing the code.

ProExpertProg replied:
I think that would add verbosity to the warnings and make the workflow more complicated: when working on passes, developers would have to manually clear the cache. This approach doesn't hurt any particular situation.

Unless you're talking about the case where outside developers supply their own passes - but there they can also just not inherit from InductorPass and the cache will always be bypassed. Or they can inherit and provide their own UUID which doesn't change and deal with caching themselves. Or they set it properly and get proper caching.

youkaichao replied:
yeah I'm talking about the case where outside developers supply their own passes. It makes sense for me to clear the cache if you change the compilation setting. and we will add per-config compilation cache later, and only use cache when users specify it (by default not using cache from previous run)

ProExpertProg committed:
Fix log message V1

Signed-off-by: luka <luka@neuralmagic.com>
Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

ProExpertProg committed:
PR comments:
- rename functionalization.py
- add detailed comment about caching

Also:
- rename TODOs
- format

Signed-off-by: luka <luka@neuralmagic.com>
Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

youkaichao commented on vllm/compilation/pass_manager.py:
```diff
@@ -0,0 +1,88 @@
+        passes_state = []
+        for pass_ in self.passes:
+            if isinstance(pass_, InductorPass):
+                passes_state.append(pass_.uuid())
```
with the discussion in the channel, I agree it is useful to cache based on the content of the pass.

however, i still think using the hash of the whole file is too overkill. you can use python `inspect` module to get the source code from the function object, which i think should be enough.

you cannot really cache based on the full code, e.g. even if the file does not change, a dependent package changes, and you cannot invalidate the cache because you don't know dependent package version changes.

youkaichao replied:
even if you want to hash the whole file, i think python's `inspect` module should give your enough information.

my request is, if a developer passes a function's qualname, it should just work. they don't need to know this uuid stuff.

ProExpertProg replied:
Okay yeah that's fair. I'll look into `python.inspect`. Should we use this for vLLM passes as well? Or just custom user passes?

ProExpertProg replied:
The nice things about the current approach with vLLM passes is that we do know the files a pass depends on. But I think inspect would be fine too.

youkaichao replied:
we can inspect both vllm passes and custom user passes.

ProExpertProg replied:
Done

ProExpertProg committed:
Add rotary embedding to func test

Signed-off-by: luka <luka@neuralmagic.com>
Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

ProExpertProg committed:
Add logging & timing to passes

Signed-off-by: luka <luka@neuralmagic.com>
Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

ProExpertProg committed:
Inspect functions for UUID, add UUID to non-managed passes, clean up, move VllmInductorPass into own file

Signed-off-by: luka <luka@neuralmagic.com>
Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

youkaichao reviewed (approved):
thanks for the great work!

youkaichao commented:
edge cases still:

<img width="870" alt="image" src="https://github.com/user-attachments/assets/0ab1be5a-526d-4585-8776-2e6aeb1e36dc">

ProExpertProg commented:
Huh, it worked for me, maybe I didn't check the final changes

ProExpertProg committed:
Fix assert

Signed-off-by: luka <luka@neuralmagic.com>
Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

ProExpertProg committed:
Fix config

Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

ProExpertProg committed:
Fix class inspection

Signed-off-by: Luka Govedic <luka.govedic@gmail.com>

youkaichao commented:
@ProExpertProg great job!
